### PR TITLE
Adds option for running plugin through babel-preset-env pipeline

### PR DESCRIPTION
**Pull request by mcMickJuice**
https://github.com/square/babel-codemod/issues/29

Points of discussion/questions:

* Not sure of naming of this cli option
* How can we write a unit test against this?

### CODE REVIEW

**Comment by eventualbuddha** [comment]
@mcMickJuice sorry it took me so long to get to this. I've mostly finished it up, leaving the option name the same as you had it. I added a test to ensure `export default …` plugins work as expected. What do you think?

**Comment by mcMickJuice** [comment]
No problem and all looks good. Ship it!

**Comment by eventualbuddha** [comment]
Cool, still some things I need to do first:

- [ ] update help docs to encourage `--babel-transform` instead of `babel-register`
- [ ] add an equivalent option for TypeScript?
- [ ] ensure plugins with multiple files get transpiled correctly

**Comment by mcMickJuice** [comment]
For typescript, should we have a separate issue to tackle that? along with doc updates and tests?

**Comment by eventualbuddha** [comment]
Yeah, I’ll make a ticket for TypeScript.

**Comment by mcMickJuice** [comment]
> update help docs to encourage --babel-transform instead of babel-register

Updated readme, explaining how to use new switch and when to use `babel-register` instead of `--babel-transpile`.  Split out example of using ts-node/register

**Comment by mcMickJuice** [comment]
> ensure plugins with multiple files get transpiled correctly

Currently working on this and my test is failing. Seems like imported modules are not being transpiled...

**Comment by eventualbuddha** [comment]
> Seems like imported modules are not being transpiled...

Yeah, I didn’t really explain it in my comment, but that’s something that I was reasonably sure wouldn’t work correctly with the way I implemented it. The problem is that the function I pass to babel-register to determine which files should be transpiled only whitelists the path that was `require.resolve`-d, but nothing imported or required by that file.

A better approach might be to do the `require.resolve`, then somehow whitelist everything in that package (if it is a package) or in its directory, possibly excluding `node_modules` and other stuff like that. Unfortunately, this is somewhat complex and I’m open to alternative suggestions. For example, what about doing the `require` inside a `try` block and, if the error is a `SyntaxError`, turn on babel-register for everything and try again?

**Comment by mcMickJuice** [comment]
Yep, removing the ignore function fixes this issue.

Are you looking to avoid transpiling node_modules for purposes of speed?  If so, [babel-register does this by default](https://babeljs.io/docs/usage/babel-register/#usage-ignores-node-modules-by-default).

**Comment by mcMickJuice** [comment]
We could also just accept an ignore regex in the event that someone is using an npm package that is not transpiled.  Essentially exposing the same config setting that babel-register accepts

**Comment by eventualbuddha** [comment]
By default babel-register also only transpiles files in packages that have babel-register as a dependency, right? I think that’s one issue I was trying to avoid eight the approach I took.

**Comment by mcMickJuice** [comment]
I'm not sure about that. But could we consider this an edge case and not alter babel-register's default behavior?  If there are future issues we can have a bug opened and add further extension points (e.g. exposing babel-register config to cli).

If you're cool with this, I can remove the whitelisting and push the test I added.

**Comment by eventualbuddha** [comment]
Sure, let’s start with this and file bugs for other issues.

**Comment by eventualbuddha** [comment]
What do you think about making this on by default?

**Comment by mcMickJuice** [comment]
Completely agree.  I just updated docs to say this functionality is on by default.  I'll move forward with turning this on by default and exposing a switch to use babel-register.

Since this project now has a dependency on babel-register, should the switch be `--use-local-babel` which bypasses the registering of babel-preset-env and just calls `babel-register` without any arguments.  If that sounds good to you I can start on that.

**Comment by mcMickJuice** [comment]
Not sure why my pushes are failing in the CI env 🤷‍♂️

**Comment by mcMickJuice** [comment]
Updated to use `--use-local-babel` switch and updated functionality to opt out of using babel-preset-env, updated tests and docs.  Couldn't think of a good way to write a test to ensure that local babelrc file was being used if `--use-local-babel` switch is passed

File: src/index.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ OPTIONS
   -o, --plugin-options PLUGIN=OPTS  JSON-encoded OPTS for PLUGIN (allows multiple).
   -r, --require PATH                Require PATH before transform (allows multiple).
       --extensions EXTS             Comma-separated extensions to process (default: "${Array.from(DEFAULT_EXTENSIONS).join(',')}").
+      --use-local-babel             Run plugins through babel plugins/presets specified in local .babelrc file instead of babel-preset-env
```
Comment (eventualbuddha) [inline src/index.ts]: nit: end this statement in a period.

File: test/fixtures/plugin/increment-export-default-multiple/increment-value.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import path from 'path'
+
+const i = path.join(__dirname, 'hello.js')
```
Comment (eventualbuddha) [inline test/fixtures/plugin/increment-export-default-multiple/increment-value.js]: Why is this here?

File: test/fixtures/plugin/increment-export-default-multiple/increment-value.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+export default function incrementValue(value) {
+    return value + 1
```
Comment (eventualbuddha) [inline test/fixtures/plugin/increment-export-default-multiple/increment-value.js]: nit: 2-space indent, use semicolons. (I should set up prettier on this project)

File: test/fixtures/plugin/increment-export-default-multiple/increment-value.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+}
```
Comment (eventualbuddha) [inline test/fixtures/plugin/increment-export-default-multiple/increment-value.js]: nit: use trailing newlines

File: test/fixtures/plugin/increment-export-default-multiple/increment-export-default.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import incrementValue from './increment-value'
+
+export default function() {
+    return {
```
Comment (eventualbuddha) [inline test/fixtures/plugin/increment-export-default-multiple/increment-export-default.js]: nit: use 2-space indent

File: test/fixtures/plugin/increment-export-default-multiple/increment-export-default.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+      visitor: {
+        NumericLiteral(path) {
+          path.node.value = incrementValue(path.node.value)
```
Comment (eventualbuddha) [inline test/fixtures/plugin/increment-export-default-multiple/increment-export-default.js]: nit: use semicolons

File: test/CLITest.ts [hunk 7]
```diff
@@ -187,4 +187,49 @@ describe('CLI', function() {
...
+      await readFile(afile, 'utf8'),
+      '4 + 5;'
+    );
+  });
+
+  it('fails when specifying --use-local-babel as there are no plugins loaded', async function() {
+    let afile = await createTemporaryFile('a-file.js', '3 + 4;');
+    let { status } = await runCodemodCLI([afile, '-p', plugin('increment-export-default'), '--use-local-babel']);
+
+    const syntaxError = 255;
```
Comment (eventualbuddha) [inline test/CLITest.ts]: I get where you’re coming from with having this as `const`, but let’s just use `let`.

File: test/CLITest.ts [hunk 8]
```diff
@@ -187,4 +187,49 @@ describe('CLI', function() {
...
+    );
+    strictEqual(
+      await readFile(afile, 'utf8'),
+      '4 + 5;'
+    );
+  });
+
+  it('fails when specifying --use-local-babel as there are no plugins loaded', async function() {
+    let afile = await createTemporaryFile('a-file.js', '3 + 4;');
+    let { status } = await runCodemodCLI([afile, '-p', plugin('increment-export-default'), '--use-local-babel']);
```
Comment (eventualbuddha) [inline test/CLITest.ts]: We should also make some sort of assertion about the output. Ideally we could even catch `SyntaxError` and provide a better error message telling users they may need to use a custom `.babelrc`.

File: src/Options.ts [hunk 9]
```diff
@@ -83,6 +92,16 @@ export default class Options {
     }
   }
 
+  loadBabelTranspile() {
+    let pluginOptions;
+    if(!this.useLocalBabel) {
```
Comment (eventualbuddha) [inline src/Options.ts]: nit: space between `if` and `(`

File: src/index.ts (hunk 1 above)
Comment (eventualbuddha) [reply src/index.ts]: Also, should we add a `--no-transpile-plugins` option? Using `babel-register` does slow things down a bit and it might be nice to allow opting out of that. What do you think?

File: README.md [hunk 10]
```diff
@@ -32,18 +32,29 @@ For more detailed options, run `codemod --help`.
...
 
-While testing out your plugin, you may find it useful to use the `--require` option when running `codemod` if your plugin is written using JavaScript syntax not supported by the current version of node. For example:
+### Transpiling using babel plugins
+
+`babel-codemod` also supports non-standard/future language features that are not currently supported by the latest version of node.  It does this by leveraging `babel-preset-env` which loads the [latest babel plugins](https://github.com/babel/babel/tree/master/packages/babel-preset-env#support-all-plugins-in-babel-that-are-considered-latest).  This feature is on by default.
+
+This feature should support most use cases when writing plugins in advanced Javascript syntax. However, if you are writing plugins with syntax that is beyond "latest", or you would like to use your own set of plugins and presets, you can pass in the `--use-local-babel` switch in combination with a local `.babelrc` file that lists the presets/plugins you want applied to your plugin code.
```
Comment (eventualbuddha) [inline README.md]: nit: capitalize as “JavaScript”

File: README.md [hunk 11]
```diff
@@ -32,18 +32,29 @@ For more detailed options, run `codemod --help`.
...
 ```
-# Run a local plugin written with newer JavaScript syntax.
-$ codemod --require babel-register --plugin ./my-plugin.js src/
+# Run a local plugin that is passed through locally installed babel plugins
+$ codemod --use-local-babel --plugin ./my-plugin.js src/
+```
+
+This requires that all babel plugins and presets be installed locally and are listed in your `.babelrc` file.  `babel-codemod` uses `babel-register` under the hood too accomplish this and all `.babelrc` [lookup rules apply](https://babeljs.io/docs/usage/babelrc/#lookup-behavior).
 
+### Transpiling using typescript
```
Comment (eventualbuddha) [inline README.md]: nit: capitalize as “TypeScript”

File: src/index.ts (hunk 1 above)
Comment (mcMickJuice) [reply src/index.ts]: is this something we can have as a separate, dependent issue on this feature? Just to keep this PR focused?

File: test/CLITest.ts (hunk 8 above)
Comment (mcMickJuice) [reply test/CLITest.ts]: It's difficult to find where to place the try catch to determine if the SyntaxError is a result of the plugin or the code that we're modding.  We could also look at the stack trace to determine where the syntax error occurred?:

```
/Users/z002s72/dev/git-clones/babel-codemod/test/fixtures/plugin/increment-export-default.js:1
(function (exports, require, module, __filename, __dirname) { export default function() {
                   ^^^^^^

SyntaxError: Unexpected token export
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:599:28)
    at loader (/Users/z002s72/dev/git-clones/babel-codemod/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/z002s72/dev/git-clones/babel-codemod/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
```

Comment (mcMickJuice) [reply test/CLITest.ts]: Looks like wrapping the following in try/catch captures the specific plugin related SyntaxError:

``` 
// index.ts, line 76
let plugins = options.getBabelPlugins();
```

However the error is undefined...im little lose there :)

Comment (eventualbuddha) [reply test/CLITest.ts]: I’ll take a look when I get home.

**Comment by eventualbuddha** [comment]
Released in v1.3.0 with the final CLI options being `--[no-]transpile-plugins` (on by default) and `--[no-]find-babel-config` (off by default). Thanks @mcMickJuice!
